### DISCUSSION TITLE
Arbitrary read > process > write functionality

### DISCUSSION THREAD

**Discussion by haz** [aig-upf/tarski#108] [Ideas]
Part of what excites me most about tarski is the promise of having an arbitrary PDDL-like tool suite at my finger-tips. If you take a look at what I did for epistemic planning, we have...

- [Include operations](https://github.com/QuMuLab/pdkb-planning/blob/master/examples/planning/grapevine-doxastic/domain-4ag.pdkbddl) to split up PDDL files
- Special syntax to [flag fluent definitions](https://github.com/QuMuLab/pdkb-planning/blob/master/examples/planning/grapevine-doxastic/domain.pdkbddl#L7)
- [Added fields](https://github.com/QuMuLab/pdkb-planning/blob/master/examples/planning/grapevine-doxastic/domain.pdkbddl#L20) for the action schema
- [Custom syntax](https://github.com/QuMuLab/pdkb-planning/blob/master/examples/planning/grapevine-doxastic/prob-4ag-8g-2d.pdkbddl#L18) for the individual fluents (wherever they may appear)

Now, I spent quite a few hours putting in parser hacks to make that viable (and it's not pretty :P). My hope would be that things like this are an order of magnitude easier with something like `tarski`. I don't mean functionality supported out of the box, but rather _the flexibility to add functionality at a whim_.

How close are we to that dream?

**Comment by miquelramirez** [comment]
Hi Chris,

Something like that motivated the development of `tarski`. One of the tutorials covers the mother of all epistemic planning approaches.

Note that `tarski` is not meant to be just a "pddl tool", as pddl is just one way to specify succinctly and in a modular way planning instances. So for instance "splitting pddl files" is as hard or as easy as loading the original files, manipulate the AST as one wants and write back as many pddl specifications as you want or need.

Parsers in my experiences tend to be very hard to extend and that is because of how tightly coupled is the interpretation of the different produced grammar rules. I am thinking that perhaps moving on from antlr and adopt `ply` would be a sensible idea, that would make easier for peeps to embark onto language design (a sorely neglected research are in planning if you ask me).

I haven't followed up the links on your post yet, I presume that I will find examples of what you mean?

**Reply by haz** [reply]
Well the links just highlight examples of the frankenstein-like syntax I used for ep-planning.

Certainly don't think `tarski` should accommodate an ever-growing language set, but rather it'd be awesome if it minimized the time from idea on a new language addition/feature to having it parse-able / useable.

I generally pointed students and researchers to https://github.com/boompig/pddl-parser-2 over the FD parser. Why? Well it reads in any arbitrary nonsense you place in the parse-tree (not checking a grammar for the right `:parameters`, `:preconditions`, `:effects` symbols on an action schema, for example).

It doesn't grant you syntax checking, but at least means it's fairly minimal to go from a new language addition (assuming it's lisp-like) to having a datastructure you can work with.

As for the ep-plan stuff (markedly not lisp-like), I opted to do a first pass on it with just python and regex, converting it into a more cumbersome looking lisp-like format for the parser to handle well.

**Reply by miquelramirez** [reply]
> it'd be awesome if it minimized the time from idea on a new language addition/feature to having it parse-able / useable.

A parser checks whether the input satisfies the grammar constraints, so something like what you suggest would require to come up with a grammar that leaves some elements to be very loosely constrained. Something like a meta-PDDL that would probably look a lot like LISP.

Another of the main reasons for investing time on `tarski` came from how bored I grew of hunting bugs and syntax errors with older codes that slightly refactored Malte's original FD front end.

**Reply by haz** [reply]
> Something like a meta-PDDL that would probably look a lot like LISP.

Ya, pretty much. But with whatever exists already in python, it's not that big of a lift.

Here's a better way to phrase the entire motivation: say I want to invent a new form of non-deterministic planning, **agent-driven uncertainty**, and I'd like the syntax to look like...

```
(:action whatever
   ...
   :effect (and (ag-oneof (?ag - agent) (some-effect) (some-other-effect)))
   ...)
```

It's not all that different from what's currently there in regular non-determinism. Special tag for `ag-oneof`, first argument is the agent, and the rest is the choice of effects.

**What steps must be taken, with `tarski`, in order to get things like that parsed and ready to manipulate?**

Idea isn't for `tarski` to adopt the `ag-oneof` syntax (it's something I just made up), but rather support the usecase for me to arbitrarily make stuff up and ship my research with tarski configurations that do the trick (rather than having a million little tarski's out there configured slightly differently). Actually, a million little tarski's sounds kinda cute...

**Reply by gfrances** [reply]
Hi Christian,

it's definitely a good use case, I've thought about this myself a few times.
I would say the current parser based on antlr is not the most user-friendly thing to extend.
Antlr has some limited notion of grammar inheritance / reusability (see e.g. https://github.com/antlr/antlr4/blob/master/doc/grammars.md#grammar-imports),
but I don't think that covers exactly all we'd need here. 

We could do some work to facilitate the use case within the current design, and document a bit the steps to be taken to extend the parser. It's not that many, e.g.: create a new `.g4` grammar file, compile it, adapt the reader (perhaps by subclassing or extending the [reader class](https://github.com/aig-upf/tarski/blob/3704b7a4c3f8cbdb8c1de8f930a7d9727751d37d/src/tarski/io/_fstrips/reader.py), etc.).
But still I'm not sure if that'd be user-friendly enough.

The alternative would be to go the route Miquel suggests, or exploit the fact that, after all, parsing LISP-like structures does not require a big deal of complexity (see e.g. the fast downward [parsing classes](https://github.com/aibasel/downward/tree/main/src/translate/pddl_parser), but then there's also a good amount of things that we lose or need to implement ourselves, like good error reporting, etc.

**Reply by haz** [reply]
Introduced functionality to cover/simplify this:

> create a new .g4 grammar file, compile it, adapt the reader (perhaps by subclassing or extending the reader class, etc.).

...might just do the trick.

```python
tarski.language.add_new_primitive(
  'oneof',
  requirement=':non-deterministic',
  valid_locations=[':effect'],
  min_args=2,
  max_args=None
)

tarski.parse_stuff(...)
```

You get the point. Essentially do all the funky antlr stuff behind the scenes, and expose functionality that covers 90% of the use-cases we have for augmenting a language.

**Reply by gfrances** [reply]
I see what you mean. This being Python, everything is very much possible, but one problem I see with this design is that it'd require the user to have the entire ANTLR installed on the system (ATM we only require the Pypi-installable `antlr4-python3-runtime` package). Of course that might be an acceptable price to pay, if you're willing to do this stuff, but might put off some of your users.

I can imagine though that if we go for something Python-native, like `ply`, as mentioned by Miquel, this might become much more feasible?

**Reply by haz** [reply]
Aha! Didn't realize that it ships with pre-compiled grammar defs. Absolutely makes sense, and this just means I'd vote 100x over for moving to `ply` for things like this. I'd prefer that route over having to require the full antlr system.

**Reply by JoanEspasa** [reply]
aye, `ply` should be the way to go.

**Reply by gfrances** [reply]
Actually, I'm seeing that Ply has been somewhat superseeded by [Sly](https://github.com/dabeaz/sly), but otherwise yes, I agree it'd be cool to give this a try. If anyone has a few spare cycles for this, I could definitely try to help on this.

**Comment by hectorpal** [comment]
# Pick and mix translations

In a recent slack discussion, @haz said:
> I'll likely put together & release a wrapper on top of tarski that lets you do common compilation things (ground with a single line, flatten to avoid objects/constants, clone original state fluents, etc).

It’d be nice to enable parametric literals for translations like K(lit, set_of_lit), and lifted versions of that.
This would enable “pick and mix” translations like:
- Lets put few softgoals, some conformant, some epistemic.
- Now let’s solve that classical problem.
- Better: solve a FOND problem.

@haz added:
- That gets at having custom PDDL syntax (like I did for epistemic planning).
- Requires some better parsing / BNF support, though...
